### PR TITLE
Introducing MKL to softmax for inference

### PR DESCRIPTION
**Pull request by jczaja**
This PR is introducing MKL based execution of softmax operator.

Capi DAM test's profiling shows **~2** times improvement in softmax op execution with this optimization.
**Num threads:** 1
**Batch:** 1,8,32,128
**Platform:** Intel(R) Xeon(R) Platinum 8180 CPU @ 2.50GHz

**Notes:**
- Optimization is enabled when Paddle is configured with: ON_INFER = ON flag
- To have unit test for it , just build with ON_INFER=ON and run test_softmax_op.py

### CODE REVIEW

File: paddle/fluid/operators/softmax_op.h [hunk 1]
```diff
@@ -35,8 +35,10 @@ class SoftmaxKernel : public framework::OpKernel<T> {
     Tensor X_2d = framework::ReshapeToMatrix(*X, rank - 1);
     Tensor Out_2d = framework::ReshapeToMatrix(*Out, rank - 1);
 
-#ifdef ON_INFER
-    math::SoftmaxFunctor<DeviceContext, T, true>()(
+#ifdef PADDLE_ON_INFERENCE
+    math::SoftmaxFunctor<
+        DeviceContext, T,
+        std::is_same<DeviceContext, platform::CPUDeviceContext>::value>()(
         context.template device_context<DeviceContext>(), &X_2d, &Out_2d);
```
Comment (tpatejko) [inline paddle/fluid/operators/softmax_op.h:42]: Why is this `std::is_same` needed? The result of this expression sets `is_test` template parameter. The expression checks if DeviceContext is of type CPUDeviceContext. The same is done by `std::enable_if` in the preceding code. One of them seems to be redundant.

File: paddle/fluid/operators/math/softmax.h [hunk 2]
```diff
@@ -19,7 +19,8 @@ namespace paddle {
 namespace operators {
 namespace math {
 
-template <typename DeviceContext, typename T, bool is_test>
+template <typename DeviceContext, typename T, bool is_test,
```
Comment (tpatejko) [inline paddle/fluid/operators/math/softmax.h:22]: Suggestion just for the future reference. You can safely ignore it. 

Using the boolean `is_test` flag makes code a bit difficult to follow due to constant values `true` or `false` in the specializations of the class.

Maybe in future you could try to experiment with property type for inference defined along these lines:
``` struct is_test {};``` and use it whenever type specific for inference is used.

File: CMakeLists.txt [hunk 3]
```diff
@@ -302,6 +302,14 @@ set(PADDLE_PYTHON_BUILD_DIR "${CMAKE_CURRENT_BINARY_DIR}/python/build")
...
 set(CMAKE_C_FLAGS_RELWITHDEBINFO "-O3 -g -DNDEBUG")
 
+if (ON_INFER)
+    message(STATUS "On inference mode, will take place some specific optimization.")
+    add_definitions(-DPADDLE_ON_INFERENCE)
+else()
+    #TODO(luotao), combine this warning with `make inference_lib_dist` command.
+    message(WARNING "On inference mode, will take place some specific optimization. Turn on the ON_INFER flag when building inference_lib only.")
+endif()
+
```
Comment (tpatejko) [inline CMakeLists.txt:312]: This code is partially present in the main CMakeLists.txt file:
https://github.com/PaddlePaddle/Paddle/blob/f4c869d872a62d99cfbbd3e3c5c5d0cf2db4d863/CMakeLists.txt#L316-L321

Is it necessary to copy just to set an additional define?

Comment (jczaja) [reply CMakeLists.txt:312]: I do not quite understand.  I moved quoted section from very bottom of main CMakeLists.txt to a bit above in the same CMakeLists.txt. I cannot see any redundancy here

File: paddle/fluid/operators/math/softmax.h (hunk 2 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax.h:22]: Interesting suggesstion will look to it closer on my next PR related to softmax for inference

File: CMakeLists.txt (hunk 3 above)
Comment (luotao1) [reply CMakeLists.txt:312]: @tpatejko These codes are only moved, not copied.

File: paddle/fluid/operators/softmax_op.h (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/softmax_op.h:42]: You are correct, that condition is left over from template experiemnts. Let me remove it in my next softmax for inference PR

File: paddle/fluid/operators/math/softmax.h [hunk 4]
```diff
@@ -19,7 +19,8 @@ namespace paddle {
...
+          typename Enable = void>
```
Comment (luotao1) [inline paddle/fluid/operators/math/softmax.h:23]: `Enable`: Could you rename this parameter name to make it much clearer?

File: paddle/fluid/operators/math/softmax_impl.h [hunk 5]
```diff
@@ -65,36 +66,46 @@ void SoftmaxFunctor<DeviceContext, T, is_test>::operator()(
...
-                                               softmax.sum(along_class)
-                                                   .inverse()
-                                                   .eval()
-                                                   .reshape(batch_by_one)
-                                                   .broadcast(one_by_class));
+    // 2D data. Batch x C
+    const int batch_size = in_dims[kBatchDim];
+    const int num_classes = in_dims[kClassDim];
+    std::vector<float> entities(batch_size);
+    auto blas = math::GetBlas<DeviceContext, float>(context);
```
Comment (luotao1) [inline paddle/fluid/operators/math/softmax_impl.h:87]: how about `int xxx= n * num_classes` outside of `for`to make less compute?

File: paddle/fluid/operators/math/softmax_impl.h [hunk 6]
```diff
@@ -65,36 +66,46 @@ void SoftmaxFunctor<DeviceContext, T, is_test>::operator()(
...
+    for (int n = 0; n < batch_size; ++n) {
+      entities[n] = in_data[n * num_classes];
+      for (int c = 1; c < num_classes; ++c) {
+        entities[n] = in_data[n * num_classes + c] > entities[n]
+                          ? in_data[n * num_classes + c]
+                          : entities[n];
```
Comment (luotao1) [inline paddle/fluid/operators/math/softmax_impl.h:93]: for line91-93, does following codes much faster?
```
if(n_data[n * num_classes + c] > entities[n]) entities[n] = in_data[n * num_classes + c]
```

Besides, line89-93 is used for finding max of `in_data[n * num_classes + c]` (c=0->num_classes), could these lines use  [cblas_i?amax ](https://software.intel.com/en-us/mkl-developer-reference-c-cblas-i-amax) function?

File: CMakeLists.txt (hunk 3 above)
Comment (tpatejko) [reply CMakeLists.txt:312]: My mistake. I haven't noticed that.

File: paddle/fluid/operators/math/softmax_impl.h [hunk 7]
```diff
@@ -65,36 +66,46 @@ void SoftmaxFunctor<DeviceContext, T, is_test>::operator()(
...
+            in_data[n * num_classes + c] - entities[n];
+      }
+    }
+
+    blas.VEXP(num_classes * batch_size, out_data, out_data);
+    for (int n = 0; n < batch_size; ++n) {
+      entities[n] = out_data[n * num_classes];
+      for (int c = 1; c < num_classes; ++c) {
+        entities[n] += out_data[n * num_classes + c];
+      }
```
Comment (luotao1) [inline paddle/fluid/operators/math/softmax_impl.h:106]: line 103-106 are used for sum `out_data[n * num_classes + c]` (c=0->num_classes), could these line use [cblas-asum](https://software.intel.com/en-us/mkl-developer-reference-c-cblas-asum)?

**Review by tpatejko — APPROVED** [review]

File: paddle/fluid/operators/math/softmax_impl.h (hunk 5 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax_impl.h:87]: From my experience, compiler can detect  such computation to be done many times on same data and put it automatically out of loop

File: paddle/fluid/operators/math/softmax_impl.h (hunk 6 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax_impl.h:93]: No, I do not think we can use that. cblas_i?amax searches for maximal **absolute** value element. And At that moment we are before exponenting array so there may be both negative and positive values. So we could get diffrent value got as maximum values chosen than in orginal code

File: paddle/fluid/operators/math/softmax_impl.h (hunk 7 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax_impl.h:106]: Yes at that situation we could do that but I'm working on next part of optimization ot it eg. enhancing vectorization using OpenMP simd directives. I will check if cblas-asum is better than openmp simd code and present faster one in next PR

**Comment by luotao1** [comment]
how do you think https://github.com/PaddlePaddle/Paddle/pull/14437#discussion_r234585222? @ jczaja

File: paddle/fluid/operators/math/softmax.h (hunk 4 above)
Comment (jczaja) [reply paddle/fluid/operators/math/softmax.h:23]: Ok, I will change a name of mentioned template param in next PR. Do you have any suggestions which name would be clearer?

Comment (luotao1) [reply paddle/fluid/operators/math/softmax.h:23]: How about rename to `enable_if_CPU`?

**Comment by luotao1** [comment]
Looking forward to your next PR!

**Pull request merged**